### PR TITLE
[Doc] #643 Tutorial for Running postgreSQL integration tests

### LINKED ISSUE #643: Document how to run postgres integration tests
We need to create a tutorial showing how to run postgres integration tests. It should cover the following:

- instructions to install dependencies (dockerctx, psycopg2, pgspecial<2)
- tips for fixing common errors

e.g., if docker isn't running, I get the following:

```
docker.errors.DockerException: Error while fetching server API version: ('Connection aborted.', FileNotFoundError(2, 'No such file or directory'))
```

- commands to run tests

Note that postgres tests are distributed in two files: `test_postgreSQL.py` and `src/tests/integration/test_generic_db_operations.py`, however, the latter contains tests for other databases so the tutorial should demonstrate how to run postgres tests and skip the rest.

Another problem I encountered is that when running postgres tests in the `test_generic_db_operations.py`, the command gets stuck for ~1 minute if I don't have psycopg2 installed. there seems to be something wrong with our tests since failing to import psycopg2 should be fairly quickly, and the command should exit within a few seconds, so we need to investigate and fix that as well.

We can put this guide in a separate ipynb file, and link from [here](https://jupysql.ploomber.io/en/latest/community/developer-guide.html#integration-tests), to the new tutorial. note that we'll have to [exclude](https://github.com/ploomber/jupysql/blob/4b938edcdea296dcf1c9585b6ed52fadf50835eb/doc/conf.py#L18) the tutorial from execution since readthedocs cannot spin up docker containers.

Take into account that you should automate as much aspects as you can. Ideally, a developer would open your tutorial,  run the commands from start to finish, and have all the postgres tests run without having to do any manual modifications to the configuration, or source code. The existing tutorials in the documentation are [good examples](https://jupysql.ploomber.io/en/latest/integrations/postgres-connect.html); however, note that the linked tutorial targets end-users (jupysql users) and this tutorial targets jupysql developers.

When opening the PR, verify that all check in the CI pass. If there are issues, fix them before requesting a review. If in doubt, you can ask for help.

To write the tutorial, you can write a notebook with the [bash-kernel](https://github.com/takluyver/bash_kernel)

*Note*: the comments on this issue are locked. If you need to ask questions, please open a draft PR, tag me and ask your questions.

### PR DESCRIPTION
**Pull request by sa4084**
## Describe your changes
This is a pr for creating a documentation to run postgreSQL integration tests. 
2 files changes:
1. Created a tutorial postgres.ipynb file.
2. Excluded it in conf.py

## Issue number

643 

[Doc] 
1. Advises users to download the requirements for the tests.
2. Identifies Problems and provides possible solutions.
3. Runs postgreSQL integration tests in both test_postgreSQL and test_generic_db_operations.py by suggesting local changes in order to run test_generic_db_operations.py file exclusively for postgreSQL tests.

## Checklist before requesting a review

- [✅] Performed a self-review of my code
- [✅] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [✅] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [✅] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--672.org.readthedocs.build/en/672/

### CODE REVIEW

**Comment by sa4084** [comment]
Hi, @edublancas

Questions
- Should I update the CHANGELOG file right now ?
- Is it within the scope that I should also update the tests in test_generic_db_operations.py with the local changes required to run as mentioned in my doc ?
Example : Adding this code snippet before each function
    if ip_with_dynamic_db != ALL_DATABASES[0]:
        pytest.skip(f"Skipping test for {ip_with_dynamic_db}")
        return

**Review by edublancas — CHANGES_REQUESTED** [review]
Some feedback:

- there's no need for users to run `docker run`. if you look at the source code, you'll see that whenever users run `pytest` a postgres container starts automatically. as long as they have the dependencies installed, it'll work
- your docs build is failing, please check the errors
- the instructions aren't very automated. asking the user to modify their `sys.path` isn't ideal, it'd be better to test whether the scripts to setup the dev environment are correctly installing the SQL package (they're, this shouldn't be a problem)
- I think you're suggesting to modify the tests by adding an if statement to the source code. this isn't ideal because it'd take a lot of time to add/remove this every time a developer wants to run/skip certain tests. is there a way that we can do something like `pytest --some-argument` and then skip all non-postgres tests? (spoiler: there is)

**Comment by sa4084** [comment]
Hi @edublancas ,

I have fixed the build issues and taken your suggestions into account as well. 
Thank you for the spoiler, because of that I was able to find a `-k` parameter in the pytest custom marker documentation to ignore the non-postgres tests.
I also cleaned the documentation and tested it on a different virtual environment and it autogenerated the docker container as you suggested.

Please let me know if you have any more feedback. Looking forward to hearing from you.

**Review by edublancas — CHANGES_REQUESTED** [review]
great! I approved the CI run, if you find any errors, please fix them.

I ran your tutorial and noticed that if I run:

```
pytest "src/tests/integration/test_generic_db_operations.py" -k "ip_with_postgreSQL"
```


without having psycopg2. the command takes too much time to run. seems like there is something wrong with our setup. ideally, the command should exit quickly (a failed import shouldn't take that much time to fail), please investigate (apologies, i forgot to put this in the original problem description)

**Comment by sa4084** [comment]
Hi @edublancas,

Two things:
- So, I realized I don't need the psycopg2-binary is not required dockerctx imports psycopg2 as a dependency automatically 
- Yes, you were correct it took 20 seconds just to say that it couldn't find psycopg2. This could have happened to any other package or any other error in general. So, I found a -x flag (usually used for developers) for pytest which basically stops the execution in case of any failure. I tested it from my end and it quickly showed me error in 1 second. I believe this solves the issue. Let me know if there is any other feedback.

Question:
- Please let me know if it's final then I will do the linking. Thank you.

**Comment by edublancas** [comment]
good, I see the doc build passes now!

according to pytest docs:

```
  -x, --exitfirst       Exit instantly on first error or failed test
```

so yeah, this can help. however, it doesnt' address the root cause. take a look at the tests config and let me know if you find anything else. there is definitely something wrong in the setup because even if we omit the  `-x` flag, the tests should exit pretty quickly if we're missing psycopg2 because importing a package is pretty fast. In my case, waiting for the tests is taking 1min:


```
pytest "src/tests/integration/test_generic_db_operations.py" -k "ip_with_postgreSQL"
```

```
FAILED src/tests/integration/test_generic_db_operations.py::test_query_count[ip_with_postgreSQL-3] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_create_table_with_indexed_df[ip_with_postgreSQL-15-15] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_active_connection_number[ip_with_postgreSQL-1] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_close_and_connect[ip_with_postgreSQL-postgreSQL] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_telemetry_execute_command_has_connection_info[ip_with_postgreSQL-postgresql-psycopg2] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlplot_histogram[ip_with_postgreSQL-histogram] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlplot_histogram[ip_with_postgreSQL-hist] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlplot_histogram[ip_with_postgreSQL-histogram-bins] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlplot_boxplot[ip_with_postgreSQL-boxplot] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlplot_boxplot[ip_with_postgreSQL-box] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlplot_boxplot[ip_with_postgreSQL-boxplot-with-horizontal] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlplot_boxplot[ip_with_postgreSQL-boxplot-with] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sql_cmd_magic_uno[ip_with_postgreSQL] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sql_cmd_magic_dos[ip_with_postgreSQL] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_profile_data_mismatch[ip_with_postgreSQL] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlcmd_tables_columns[ip_with_postgreSQL-numbers] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sqlcmd_tables[ip_with_postgreSQL] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sql_query[ip_with_postgreSQL-simple-query] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sql_query[ip_with_postgreSQL-cte] - ModuleNotFoundError: No module named 'psycopg2'
FAILED src/tests/integration/test_generic_db_operations.py::test_sql_query[ip_with_postgreSQL-interpolation-like-comment] - ModuleNotFoundError: No module named 'psycopg2'
============================================================= 20 failed, 1 skipped, 130 deselected in 62.89s (0:01:02) ==============================================================

```

**Comment by sa4084** [comment]
Hi @edublancas,

I would like to share about the things I found through my investigation and some questions related to it to get more clarity.
Three Questions:

1. I actually went through the conftest.py file as you suggested. In terms of common operations, I saw that all it does it build up a connection to the database and gets url each time. 
So, my question is will not importing something that's required prevent a DB connection ? Also is it relevant to solving this problem ?

2. I saw a timeout in the db ready function but if we try to maybe reduce it we may actually be timing out when a connection could have been acquired if we waited a little longer ?

3. Is it within the scope of the problem where I have to edit anything in conftest.py file like some checks ?

File: doc/integrations/postgres.ipynb [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+   "source": [
+    "pytest -x src/tests/integration/test_generic_db_operations.py -k \"ip_with_postgreSQL\""
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "08888580",
+   "metadata": {},
+   "source": [
+    "```{tip} If you have issues [message us on Slack.](https://ploomber.io/community) ```"
```
Comment (neelasha23) [inline doc/integrations/postgres.ipynb:208]: This is not displayed correctly.
Screenshot 2023-06-29 at 11 55 42 AM

File: doc/integrations/postgres.ipynb [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "9821bbd0",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "conda create --name ploomber -y"
```
Comment (neelasha23) [inline doc/integrations/postgres.ipynb:32]: Do we need these steps? We can use the `jupysql` environment

File: doc/integrations/postgres.ipynb [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    "If you encounter Docker-related issues due to old versioning, it is recommended to [install latest Docker.](https://docs.docker.com/engine/install/) using the official installer.\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "21c6b60e",
+   "metadata": {},
+   "source": [
+    "- `pgspecial` package provides enhanced functionality for working with PostgreSQL databases, such as improved command-line interface (CLI) support etc.\n",
```
Comment (neelasha23) [inline doc/integrations/postgres.ipynb:87]: these two statements can be combined as a note

File: doc/integrations/postgres.ipynb [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+   "source": [
+    "pytest -x src/tests/integration/test_postgreSQL.py  "
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "944c80e1",
+   "metadata": {},
+   "source": [
+    "Verify that postgres is running and is ready to accept new connections in the docker terminal. After this if everything goes well it will show 3 passed testcases. "
```
Comment (neelasha23) [inline doc/integrations/postgres.ipynb:132]: Can add a screenshot here of the docker dashboard

File: doc/integrations/postgres.ipynb [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "74fb3ae2",
+   "metadata": {},
+   "source": [
+    "### Possible Errors"
```
Comment (neelasha23) [inline doc/integrations/postgres.ipynb:140]: Please format this section properly. Right now it's not very readable. Highlight important terms also, e.g., `docker.errors.DockerException`

**Comment by neelasha23** [comment]
I see this error:
Screenshot 2023-06-29 at 12 55 34 PM

File: doc/integrations/postgres.ipynb (hunk 2 above)
Comment (sa4084) [reply doc/integrations/postgres.ipynb]: Yes, we can use jupysql itself. Updated the doc.

File: doc/integrations/postgres.ipynb (hunk 1 above)
Comment (sa4084) [reply doc/integrations/postgres.ipynb:198]: Fixed.

File: doc/integrations/postgres.ipynb (hunk 5 above)
Comment (sa4084) [reply doc/integrations/postgres.ipynb:137]: Yes, I fixed them. Please check it now.

File: doc/integrations/postgres.ipynb (hunk 4 above)
Comment (sa4084) [reply doc/integrations/postgres.ipynb]: Fixed.

File: doc/integrations/postgres.ipynb (hunk 3 above)
Comment (sa4084) [reply doc/integrations/postgres.ipynb]: Fixed.

**Pull request closed without merging**